### PR TITLE
feat: Support file paths for certificate, key and client CA in web TL…

### PR DESCRIPTION
**Pull request by yp969803**
…S configuration (#6246)

## Description
fixes: #6246 
 Support file paths for certificate, key and client CA in web TLS configuration

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification
using unit tests

## Changelog entry

```Support file paths for certificate, key and client CA in web TLS configuration

```

### CODE REVIEW

**Comment by yp969803** [comment]
@simonpasquier can you review the pr, and see for any changes, after the review I will add unit tests also

**Comment by haanhvu** [comment]
Reading your code, I see that you've done two things:

- Add new fields for the paths in the CRDs
- Make sure users don't set CA/cert/key both directly and through files, only once

This is a good start, but not sufficient. You would need to extract the CA/cert/key from the files and actually make use of them. And ideally, watch those files for changes that need to be reloaded. The changes can happen in both the paths and the contents of the files.

**Comment by yp969803** [comment]
@haanhvu  "You would need to extract the CA/cert/key from the files and actually make use of them." why we do this the files are already present in volume just we have to give file path to the server so that the server can use it. I have done this. The changes will happen only if prometheus cr will be updated and if it gets updated the file path to server will automatically updated. @simonpasquier  any suggestions ?

**Comment by haanhvu** [comment]
> the files are already present in volume
Yeah I just read the issue and the code more carefully. If the requirement is the same as service monitor's endpoint's TLS config then this is probably sufficient.

**Review by simonpasquier — COMMENTED** [review]
We'd need extensive unit tests to verify all possible combinations.

File: pkg/apis/monitoring/v1/types.go [hunk 1]
```diff
@@ -310,19 +310,42 @@ type WebHTTPHeaders struct {
 	StrictTransportSecurity string `json:"strictTransportSecurity,omitempty"`
 }
 
+// ValidateSafeTLS semantically validates the safe TLS options of WebTLSConfig.
+func (c *WebTLSConfig) ValidateSafeTLS() error {
```
[the remark this answers is not in the document]
Comment (yp969803) [reply pkg/apis/monitoring/v1/types.go:314]: yep, i will remove it

**Comment by yp969803** [comment]
I will complete the unit tests as soon as possible

**Comment by yp969803** [comment]
@simonpasquier added the unit tests

File: pkg/webconfig/config.go [hunk 2]
```diff
@@ -139,11 +134,15 @@ func (c Config) addTLSServerConfigToYaml(cfg yaml.MapSlice) yaml.MapSlice {
 	}
 
 	tlsServerConfig := yaml.MapSlice{}
-	if certPath := c.tlsCredentials.getCertMountPath(); certPath != "" {
+	if certFile := c.tlsCredentials.certFile; certFile != "" {
+		tlsServerConfig = append(tlsServerConfig, yaml.MapItem{Key: "cert_file", Value: certFile})
+	} else if certPath := c.tlsCredentials.getCertMountPath(); certPath != "" {
 		tlsServerConfig = append(tlsServerConfig, yaml.MapItem{Key: "cert_file", Value: fmt.Sprintf("%s/%s", certPath, c.tlsCredentials.getCertFilename())})
 	}
```
Comment (simonpasquier) [inline pkg/webconfig/config.go:141]: (suggestion) switch/case is easier to read
```suggestion
	switch {
	case c.tlsCredentials.certFile != "":
		tlsServerConfig = append(tlsServerConfig, yaml.MapItem{Key: "cert_file", Value: c.tlsCredentials.certFile})
	case c.tlsCredentials.getCertMountPath() != "":
		tlsServerConfig = append(tlsServerConfig, yaml.MapItem{Key: "cert_file", Value: fmt.Sprintf("%s/%s", certPath, c.tlsCredentials.getCertFilename())})
	}
```

File: pkg/webconfig/config.go [hunk 3]
```diff
@@ -139,11 +134,15 @@ func (c Config) addTLSServerConfigToYaml(cfg yaml.MapSlice) yaml.MapSlice {
...
 
-	if keyPath := c.tlsCredentials.getKeyMountPath(); keyPath != "" {
+	if keyFile := c.tlsCredentials.keyFile; keyFile != "" {
```
Comment (simonpasquier) [inline pkg/webconfig/config.go:143]: same here

File: pkg/webconfig/config.go [hunk 4]
```diff
@@ -154,7 +153,9 @@ func (c Config) addTLSServerConfigToYaml(cfg yaml.MapSlice) yaml.MapSlice {
 		})
 	}
 
-	if caPath := c.tlsCredentials.getCAMountPath(); caPath != "" {
+	if caFile := c.tlsCredentials.clientCAFile; caFile != "" {
```
Comment (simonpasquier) [inline pkg/webconfig/config.go:156]: same here

**Comment by yp969803** [comment]
@simonpasquier done the above changes

**Comment by yp969803** [comment]
@simonpasquier done the changes

**Review by simonpasquier — APPROVED** [review]
Thanks!

**Pull request merged**